File: My Projects/C++/AddressBook/Address Book/AddressBook.cc
//                                    ::HARE KRISHNA::
// 										ADDRESS BOOK
//NAME    :: ASHISH DONERIYA
//COLLEGE ::MANIT
//INSPIRE BHOPAL
#include<iostream>						//for cin and cout
#include<fstream>						//for file handling	
#include<string.h>						//for string manipulation
#include<cctype>
#include<stdlib.h>						//for system("clear")
#include<math.h>						//for math functions
#include "fileutil.cc"
using namespace std;
class AddressBook;
class AddressBook:public fileutil
{
public:
	void createfile();
	void flush();						//It clears the buffer	
	void get();							//It accepts the name and phone no. from user and save it in object of class AddressBook
	void show();						//It shows the data inside 'name' and 'phone' variables of object of class AddressBook
	void advance_search();
	void extreme_search();	
	void addition();					//It is used for adding the record in the file
	void display();						//It displays all the record of the file
	void modify();						//It is used to modify an existing record in the file
	void search();						//It is used to search and display a particular record of file
	void deletion();					//It is used to delete a particular record form the file
	void menu_screen();					//It displays the list of tasks that can be done on a file or it displays the options for address book
	void welcome_screen();				//It displays the welcome screen
};
void AddressBook::createfile()
{
	cout<<"Enter the file name : ";
	cin.getline(directory,30);
	fstream obj;
	obj.open(directory,ios::app);
	obj.close();
}
void AddressBook::extreme_search()
{
	ifstream obj;
	int i=0,j=0,flag,len,count;
	char str[30],**str1;
	obj.open(directory);
	if(!obj)
	{
		cout<<"File can't be opened";
		getchar();
		return;
	}
	system("clear");
	cout<<"Enter name or phone no. : ";
	cin.getline(str,30);
	len=strlen(str);
	if(len==0)
		return;
	if(check_number(str)==1&&len<=10)
	{
		for(i=0,count=0;i<total_record()-1;i++)
		{
			obj.read((char *)this,size());
			if(strcmp_adv(phone,get_phone(str),len)==1)
			{
				show();
				count=count+1;
			}
		}
		if(count==0)
			cout<<"No record found"<<endl;
	}
	else
	{
		for(i=0,count=0;i<total_record()-1;i++)
		{
			obj.read((char *)this,size());
			str1=get_combinations(str);
			for(j=0;j<4;j++)
			{
				if(strcmp_adv(name,str1[j])==1)
				{
					show();
					count=count+1;
				}
			}
		}
		if(count==0)
			cout<<"No record found"<<endl;
		else
			cout<<"\n"<<count<<" records found"<<endl;
	}
	//flush();
	getchar();
	obj.close();
}
void AddressBook::advance_search()
{
	ifstream obj;
	int i=0,flag,len,count;
	char str[30];
	obj.open(directory);
	if(!obj)
	{
		cout<<"File can't be opened";
		getchar();
		return;
	}
	system("clear");
	cout<<"Enter name or phone no. : ";
	cin.getline(str,30);
	len=strlen(str);
	if(len==0)
	{
		cout<<"No record found";
		getchar();
		return;
	}
	if(check_number(str)==1&&len<=10)
	{
		for(i=0,count=0;i<total_record()-1;i++)
		{
			obj.read((char *)this,size());
			if((long)(phone/(pow(10,10-len)))==get_phone(str))
			{
				show();
				count=count+1;
			}
		}
		if(count==0)
			cout<<"No record found"<<endl;
	}
	else
	{
		for(i=0,count=0;i<total_record()-1;i++)
		{
			obj.read((char *)this,size());
			flag=strcomp(str,name);
			if(flag==1||flag==0)
			{
				show();
				count=count+1;
			}
		}
		if(count==0)
			cout<<"No record found"<<endl;
		else
			cout<<"\n"<<count<<" records found"<<endl;
	}
	//flush();
	getchar();
	obj.close();
}
void AddressBook::flush()
{
	scanf ( "%*[^\n]" );
	scanf ( "%*c" );
}
void AddressBook::get()
{
	char ch,temp[10];
	cout<<"Phone Number : ";
	cin>>temp;
	if(!valid_num(temp))
	{
		phone=0;
		return;
	}
	else
		phone=get_phone(temp);
	cin.get(ch);
	cout<<"ENTER NAME : ";
    cin.getline(name,20);
    strcpy(name,Upper_Lower(name));
}
void AddressBook::show()
{
	cout<<"Name  :: "<<name<<endl;
	cout<<"Phone :: "<<phone<<"\n"<<endl;
}
void AddressBook::addition()
{
	ifstream in;
	ofstream out;
	int rcd=copy_into_temp();
	int i=0;
	long phone1,phone2;
	char name1[30],name2[30];
	if(	rcd==-1)
		return;
	get();
	if(!phone)
	{
		cout<<"Invalid number"<<endl;
		flush();
		getchar();
		return;
	}
	strcpy(name1,name);
	phone1=phone;
	
	in.open("temp");
	out.open(directory);
	if (!in||!out)
	{
		cout<<"Can't add the record"<<endl;
		getchar();
		return;
	}
	int flag=1;
	//while(1)
	for(i=0;i<rcd-1;i++)
	{
		in.read((char*)this,size());
		//if(in.eof()!=0)
			//break;
		if((strcmp(name1,name)<(-1))&&flag==1)
		{	
			strcpy(name2,name);
			phone2=phone;
			strcpy(name,name1);
			phone=phone1;
			out.write((char *)this,size());
			strcpy(name,name2);
			phone=phone2;
			flag=0;
		}
		out.write((char*)this,size());			
	}
	if(flag==1)
	{
		strcpy(name,name1);
		phone=phone1;
		out.write((char *)this,size());
		flag=0;
	}
	
	in.close();
	out.close();
}
void AddressBook::display()
{
	int i=1;
	ifstream obj;
	system("clear");
	obj.open(directory);
	if(!obj)
	{
		cout<<"File can't be opened";
		getchar();
		return;
	}
	while(1)
	{
		obj.read((char *)this,size());
		if(obj.eof()!=0)
			break;
		show();
		if((i++)%5==0)
		{
			getchar();
			system("clear");
		}
	}
	flush();
	obj.close();
}
void AddressBook::search()
{
	ifstream obj;
	int flag,ch,pos;
	char str[30];
	obj.open(directory);
	if(!obj)
	{
		cout<<"File can't be opened";
		getchar();
		return;
	}
	system("clear");
	cout<<"Enter name or phone no. : ";
	cin.getline(str,30);
	if(strlen(str)==0)
		return;
	if(valid_num(str)==1)
	{
		long ph=get_phone(str);
		pos=get_record_pos(ph);
		if(pos!=0)
		{
			obj.seekg((pos-1)*size(),ios::beg);
			obj.read((char *)this,size());
			show();
			getchar();
		}
		else
		{
			cout<<ph<<" is not found in Address Book"<<endl;
			char c[1];
			cin.getline(c,1);
			getchar();
			return;
		}
	}
	else
	{
		pos=get_record_pos(str);
		if(pos!=-1)
		{
			obj.seekg((pos-1)*size(),ios::beg);
			obj.read((char *)this,size());
			show();
		}
		else
		{	
			cout<<str<<" is not found in Address Book"<<endl;
			return;
		}
	}
	getchar();
	obj.close();
}
void AddressBook::modify()
{
	char str[30],ch,name1[30];
	int pos;
	long phone1;
	system("clear");
	cout<<"Enter Name or Phone Number : ";
	cin.getline(str,30);
	if(valid_num(str)==1)
		pos=get_record_pos(get_phone(str));
	else
		pos=get_record_pos(str);
	if(pos==0)
	{
		cout<<str<<" is not found in Address Book"<<endl;
	}
	else
	{
		load_record(pos);
		strcpy(name1,name);
		phone1=phone;
		cout<<name<<endl;
		getchar();
		char temp[30];
		cout<<"Enter new name : ";
		cin.getline(temp,30);
		if(strlen(temp)!=0)
			strcpy(name1,temp);
		cout<<phone<<endl;
		cout<<"Enter new phone number : ";
		cin.getline(temp,30);
		if(strlen(temp)!=0)
		{
			if(valid_num(temp)==1)
				phone1=get_phone(temp);
			else
				cout<<"Invalid Phone no."<<endl;	
		}
		replace(pos,name1,phone1);
	}
	getchar();
}
void AddressBook::deletion()
{	
	char str[30];
	int flag;
	cout<<"Enter name or number to be deleted : ";
	cin.getline(str,30);
	if(strlen(str)==0)
		return;
	if(valid_num(str)==1)
		flag=del(get_phone(str));
	else
		flag=del(str);
	if(flag==0)
		cout<<str<<" Can't be deleted"<<endl;
	//else
		//cout<<str<<" deleted"<<endl;
	flush();
}
void AddressBook::menu_screen()
{
	char ch;
	do
	{
		system("clear");
		cout<<"		          :: ADDRESS BOOK::"<<endl;
		cout<<"A::Add"<<endl;
		cout<<"V::View"<<endl;
		cout<<"S::Search"<<endl;
		cout<<"s::Advance Search"<<endl;
		cout<<"e::Extreme Search"<<endl;
		cout<<"M::Modify"<<endl;
		cout<<"D::Delete"<<endl;
		cout<<"F::Format"<<endl;
		cout<<"G::Goto Another file"<<endl;
		cout<<"E::Exit"<<endl;
		cout<<"Enter your choice : "<<endl;
		char str[30];
		cin.getline(str,30);
	    if(strlen(str)>1)
			continue;
		else
		ch=str[0];
		switch (ch)
		{
			case 'A':
			case 'a':
				addition();
				break;
			case 'V':
			case 'v':
				display();
				break;
			case 's':
				advance_search();
				break;
			case 'S':
				search();
				break;
			case 'M':
			case 'm':
				modify();
				break;
			case 'D':
			case 'd':
				deletion();
				break;
			case 'F':
			case 'f':
			{
					ofstream obj;
					obj.open(directory);
					obj.close();
					cout<<"Formatted";
					//flush();
					getchar();
					break;
			}
			case 'e':
				extreme_search();
				break;
			case 'G':
			case 'g':
				createfile();
				break;
			case 'E':
				return;
			default:
			{
						cout<<"Wrong Choice"<<endl;
			}
		}
	}while(ch!='E'||ch!='e');
}
void AddressBook::welcome_screen()
{
	system("clear");
	cout<<"+++++++++++++++++++++++++++++++++++++++++++++++++++++++++"<<endl;
	cout<<"		           ADDRESS BOOK"<<endl;
	cout<<"+++++++++++++++++++++++++++++++++++++++++++++++++++++++++"<<endl;
	cout<<"\n\n\n\nPress any key to continue................";
	getchar();
}
int main()
{
	AddressBook T;
	T.createfile();
	//T.welcome_screen();
	//T.init();
	T.menu_screen();
	return 0;
}

File: My Projects/Java/AddressBook.java
import Address_book.StringUtils;
import java.io.*;
import java.util.*;
import static java.lang.System.out;

class AddressBook extends StringUtils{
	private long mobile,landline;
	private String name,email;
	private AddressBook prev,next;
	private static String directory;
	public AddressBook(String S){
		directory=S;
		mobile=landline=0;
		name=email="";
		prev=next=null;
	}
	public AddressBook(AddressBook T){
		this.name=T.name;
		this.email=T.email;
		this.mobile=T.mobile;
		this.landline=T.landline;
		this.prev=T.prev;
		this.next=T.next;
	}
	public AddressBook(){
		initialize();
	}
	public void getCopy(AddressBook T){
		this.name=T.name;
		this.email=T.email;
		this.mobile=T.mobile;
		this.landline=T.landline;
		this.prev=T.prev;
		this.next=T.next;
	}
	public void addAtFirst(AddressBook T) throws Exception{
		AddressBook temp=new AddressBook();
		temp=T;
		temp.prev=null;
		if(this.isEmpty()==true){
			this.getCopy(T);
			this.next=this.prev=null;
		}
		else{
			temp.next=this;
			this.prev=temp;
			this.getCopy(temp);
		}
	}
	public boolean isEmpty(){
		if (this.next==null&&this.prev==null&&this.name.length()==0)
			return true;
		else
			return false;
	}
	public void delAtFirst()throws Exception{
		if(this.isEmpty()==true)
			throw (new Exception("File is Empty"));
		AddressBook temp=new AddressBook();
		if(this.next==null)
			this.initialize();
		else
			this.getCopy(this.next);
			this.prev=null;
	}
	public void delete(String  name)throws Exception{
		if(this.isEmpty()==true)
			throw (new Exception("File is Empty"));
		if ((this.name).equalsIgnoreCase(name)==true){
			delAtFirst();
			return;
		}
		AddressBook temp=this;
		AddressBook prev=temp;
		while((temp.name).equalsIgnoreCase(name)==false&&temp.next!=null){
			temp=temp.next;
		}
		if ((temp.name).equalsIgnoreCase(name)==true){
			if(temp.next==null)
				temp.prev.next=null;
			else
				temp.prev.next=temp.next;
		}
		throw (new Exception("Record Not Found"));
	}
	public void delete(long mobile)throws Exception{
		if(this.isEmpty()==true)
			throw (new Exception("File is Empty"));
		if (this.mobile==mobile){
			delAtFirst();
			return;
		}
		AddressBook temp=this;
		AddressBook prev=temp;
		while(this.mobile==mobile&&temp.next!=null){
			temp=temp.next;
		}
		if (this.mobile==mobile){
			if(temp.next==null)
				temp.prev.next=null;
			else
				temp.prev.next=temp.next;
		}
		throw (new Exception("Record Not Found"));
	}
	
	public void add(AddressBook T)throws Exception{
		if(this.isEmpty()==true){
			addAtFirst(T);
			return;
		}
		AddressBook temp=this;
		AddressBook prev=temp;
		while((Strcmp(T.name,temp.name)!=(-1))&&temp.next!=null){
			temp=temp.next;
		}
		if (Strcmp(T.name,temp.name)==(-1)){
			
			T.prev=temp.prev;
			temp.prev.next=T;
		}
	}

	public void addition(){
		try{
			get();
			AddressBook T=this;
			addRecord(T);
		}
		catch(Exception e){
			out.println(e.getMessage());
		}
	}
	public void display(){
		try{
			AddressBook T=this;
			if(this.isEmpty()==true){
				out.println("No Reord Empty");
				return;
			}
			while(T!=null){
				T.show();
			}
		}
		catch(Exception e){
			out.println(e.getMessage());
		}
	}
	public void modify(){
	}
	public void deletion(){
		out.println("Enter the name or number to be deleted : ");
		String str=(new Scanner(System.in)).next();
		if(str.length()==0)
			return;
		long mobile=0;
		try{
			if(IsValidMobile(str)==true)
				delete(Long.parseLong(str));
			else
				delete(UpperLower(str));
		}
		catch(Exception e){
			out.println(e.getMessage());
		}
	}
	
	public void search(){
		if(this.isEmpty()==true){
			out.println("NO record");
			return;
		}
		out.println("Enter the name or number : ");
		String str=(new Scanner(System.in)).next();
		String ss;
		AddressBook temp=this;
		if(IsValidMobile(str)==true){
			while(temp!=null){
				ss=(valueOf(temp.mobile));
				if(ss.indexOf(str)!=-1){
					temp.show();
				}
			}
		}
	}
	public void createFile(){
	}
	public void initialize(){
		mobile=landline=0;
		name="";
		email="";
		next=prev=null;
	}
	public void get() throws Exception{
		initialize();
		String S;
		Scanner kb=new Scanner(System.in);
		out.print("Enter name : ");
		name=kb.nextLine();
		if(name.length()==0)
			throw (new Exception("Invalid Name"));
		out.print("Enter Mobile No. : ");
		S=kb.nextLine();
		if(S.length()!=0){
			if(IsValidMobile(S)==false)
				throw (new Exception("Invalid Mobile No."));
			else
				mobile=Long.parseLong(S);
		}
		out.print("Enter Landline No. : ");
		S=kb.nextLine();
		if(S.length()!=0){
			if(IsValidNumber(S)==false)
				throw (new Exception("Invalid Landline No."));
			else
				landline=Long.parseLong(S);
		}
		out.print("Enter your email : ");
		email=kb.nextLine();
		if(email.length()!=0&&IsValidEmail(email)==false)
				throw (new Exception("Invalid Email"));
	}
	public void show()throws Exception{
		if(this.isEmpty()==true)
			throw (new Exception("Empty Record"));
		out.println("Name\t"+this.name);
		out.println("Mobile No.\t"+this.mobile);
		out.println("Landline\t"+this.landline);
		out.println("Email\n"+this.email);
	}
	public void menu(){
		String S;
		Scanner kb=new Scanner(System.in);	
		do{
			out.println("::ADDRESS BOOK::\n");
			out.println("A::Add");
			out.println("V::View");
			out.println("S::Search");
			out.println("M::Modify");
			out.println("D::Deletion");
			out.println("R::Remove File");
			out.println("F::Format File");
			out.println("G::Goto Another File");
			out.println("E::Exit");
			out.println("Enter your choice");
			S=kb.next();			
			switch (S.charAt(0)){
				case 'A':
				case 'a':
					addition();
					break;
				case 'V':
				case 'v':
					display();
					break;
				case 'M':
				case 'm':
					modify();
					break;
				case 'D':
				case 'd':
					deletion();
					break;
				case 'F':
				case 'f':{
					/*	ofstream obj;
						obj.open(directory);
						obj.close();
						cout<<"Formatted";
						*/break;
						}
				case 'S':
				case 's':
					search();
				break;
				case 'G':
				case 'g':
					createFile();
					break;
				case 'E':
					System.exit(0);
				case 'r':
				case 'R':
					out.print("Enter the file name : ");
					String temp=kb.nextLine();
					break;
				default:
					out.println("Wrong Choice");
			}
		}while(S.equalsIgnoreCase("e")==false);
	}
}
class UseAddressBook{
	public static void main(String args[]){
		AddressBook a=new AddressBook("ashish");
		a.initialize();
		try{
		}
		catch(Exception e){
			out.println(e);
		}
	}
}
